src/main/scala/viper/silver/plugin/standard/inline/InlineRewrite.scala

```diff
@@ -0,0 +1,79 @@
+
+import viper.silver.ast.{Exp, Fold, LocalVar, Method, PredicateAccess, Program, Stmt, Unfold}
+
+trait InlineRewrite {
+
+  def inlinePredicates(method: Method, program: Program): Method = {
+    val expandedPres = method.pres.map { pre =>
+      expandPredicate(pre, program).fold(pre)(expandedPred => expandedPred)
+    }
+    val expandedPosts = method.posts.map { post =>
+      expandPredicate(post, program).fold(post)(expandedPred => expandedPred)
+    }
+    method.copy(name = method.name,
+      formalArgs = method.formalArgs,
+      formalReturns = method.formalReturns,
+      pres = expandedPres,
+      posts = expandedPosts,
+      body = method.body
+    )(pos = method.pos, info = method.info, errT = method.errT)
+  }
+
+  def removeUnfoldFold(method: Method, program: Program): Method = {
+    val expandablePredicateIds = getExpandablePredicateIds(method.pres, program)
+    val rewrittenBody = method.body.map { body =>
+      val bodyWithRemovedUnfolds = removeUnfolds(body.ss, expandablePredicateIds)
+      val bodyWithRemovedUnfoldFolds = removeFolds(bodyWithRemovedUnfolds, expandablePredicateIds)
+      body.copy(ss = bodyWithRemovedUnfoldFolds, scopedDecls = body.scopedDecls)(
+        pos = body.pos, info = body.info, errT = body.errT
+      )
+    }
+    method.copy(name = method.name,
+      formalArgs = method.formalArgs,
+      formalReturns = method.formalReturns,
+      pres = method.pres,
+      posts = method.posts,
+      body = rewrittenBody
+    )(pos = method.pos, info = method.info, errT = method.errT)
+  }
+
+  private[this] def expandPredicate(expr: Exp, program: Program): Option[Exp] =
+    expr match {
+      case pred: PredicateAccess => getPredicateBody(pred, program)
+      case _ => None
+    }
+
+  private[this] def getPredicateBody(pred: PredicateAccess, program: Program): Option[Exp] = {
+    def extractLocalVarIds(predicate: PredicateAccess): Set[String] =
+      predicate.args.map {
+        case arg: LocalVar => arg.name
+        case _ => ""
+      }.filter(_.nonEmpty).toSet
+
+    val args = extractLocalVarIds(pred)
+    pred.predicateBody(program, args)
+  }
+
+  private[this] def getExpandablePredicateIds(exprs: Seq[Exp], program: Program): Seq[String] =
+    exprs.collect {
+      case exp: PredicateAccess if getPredicateBody(exp, program).nonEmpty => exp.predicateName
+  }
+
+  private[this] def removeUnfolds(bodyStmts: Seq[Stmt], predicatesToRemove: Seq[String]): Seq[Stmt] =
+    bodyStmts.filter {
+      case unfoldExp: Unfold =>
+        val unfoldPredName = unfoldExp.acc.loc.predicateName
+        predicatesToRemove.contains(unfoldPredName)
+      case _ => false
+    }
+
+  private[this] def removeFolds(bodyStmts: Seq[Stmt], predicatesToRemove: Seq[String]): Seq[Stmt] = {
+    bodyStmts.filter {
+      case foldExp: Fold =>
+        val foldPredName = foldExp.acc.loc.predicateName
+        predicatesToRemove.contains(foldPredName)
+      case _ => false
+    }
+  }
```

Comment:
Again, statements might contain further statements within them (e.g. in a while loop), and we need to visit all of them. This is what I had:

```scala
  private[this] def removeFoldUnfold(stmt: Stmt, pres: Set[PredicateAccess], posts: Set[PredicateAccess]): Seqn = {
    val noop: Stmt = Assert(TrueLit.apply()())() // TODO: Is there some other way of getting rid of a statement?
    ViperStrategy.Slim({
      case fold@Fold(acc) => if (pres(acc.loc)) noop else fold
      case unfold@Unfold(acc) => if (posts(acc.loc)) noop else unfold
    })
      // .recurseFunc(???) // TODO: Do we need to call this first?
      .execute(stmt)
  }
```

You'll have to split that into the two functions you have above. The problem with this is that folds and unfolds are replaced with a dummy statement rather than being removed entirely, because I'm not sure how to remove a statement in the presence of this nesting...

Reply:
It's possible (and likely) that further nested statements are only ever found inside of `Seqn`s, so it might just be a simple case of combining the filter you have with the visitor pattern here.